### ISSUE TITLE
Can't build .NET Core 2.1 application with Serilog 3.1.0

### ISSUE THREAD

**Issue by andrewlock** [serilog/serilog#1983]
**Description**
If you add Serilog 3.1.0+ plus to a .NET Core 2.1 application and attempt to build, you will get an error. For example, from our CI:

`/project/packages/system.runtime.compilerservices.unsafe/6.0.0/buildTransitive/netcoreapp2.0/System.Runtime.CompilerServices.Unsafe.targets(4,5): error : System.Runtime.CompilerServices.Unsafe doesn't support netcoreapp2.1. Consider updating your TargetFramework to netcoreapp3.1 or later.`

**Reproduction**
Use 3.1.0+ in a .NET ore 2.1 application

**Expected behavior**
The application should build without warnings

**Relevant package, tooling and runtime versions**
Serilog 3.1.0
.NET Core 2.1

**Additional context**

I wanted to clarify (warn?), that as Serilog 3.1.0 added a reference to _System.Diagnostics.DiagnosticSource_, which references _[System.Runtime.CompilerServices.Unsafe](https://www.nuget.org/packages/System.Runtime.CompilerServices.Unsafe/)_, you have effectively dropped support for .NET Core 2.x, due to the way Microsoft started adding `.targets` files that explicitly error when the netstandard2.0 parts are used on < .NET Core 3.1 TFMs:

If this was added in 3.0.0 it wouldn't have been an issue, but I would argue that (inadvertently) dropping TFMs in a minor version bump is very non-semver 🙈 

If you're not aware of the issue, I wrote about it last year: https://andrewlock.net/stop-lying-about-netstandard-2-support/. The onus is firmly on MS on this one IMO, but library authors don't seem to realise the implications of adding dependencies on some of these libraries (i.e. _System.Diagnostics.DiagnosticSource_). 

> For what it's worth StackExchange.Redis similarly added a dependency recently and accidentally dropped support, [as described here](https://github.com/StackExchange/StackExchange.Redis/issues/2589#issuecomment-1801901894). 

Just to be clear, I'm not at-all saying you _should_ support .NET Core 2.1 just that it _looks_ like you do to NuGet, because of the `netstandard2.0` TFM. But most of the runtimes that would _use_ this TFM (.NET Core 2.x and 3.0) won't work. Basically only Xamarin can use it without error AFAIK. 

And it's misleading in NuGet because it looks like you _should_ be able to use it without issues on .NET Core 2.x. So just consider this a heads up 😄

**Comment by nblumhardt** [comment]
Argh! Thanks for the reminder about this, @andrewlock :-)

Unfortunately it's a ship that's sailed; I think our best bet is just to document this in the release notes, and move on.

Since the _s.r.cs.Unsafe_ change, consumers on .NET Core 2.0 and 3.0 will have been locked out of package updates pretty much across the ecosystem, so in practice I don't think people still maintaining code on those frameworks will be very surprised by this, major version bump or not.

(I don't think dropping `netstandard2.0` from the package's TFMs would be a good idea, for the same reason Microsoft probably took this route in the first place: many downstream consumers will still be targeting (only) `netstandard2.0` as a way to reach a wide range supported frameworks, so in order to stop lying about compatibility for obscure and unsupported frameworks, we'd break effective compatibility with all the supported ones.)

I appreciate the heads-up, though. It's nice to have this write-up for anyone who might be searching for an explanation of that error message :+1:

Making the release notes change now.

**Comment by andrewlock** [comment]
I totally agree with you that this is going to become increasingly common, and agree the documentation approach is probably the best we can do!

Also agree that dropping netstandard2.0 is probably a non-starter. Interestingly the main reason Microsoft gave for taking the approach is analysers/source generators that _have_ to be netstandard! 

As my post probably makes clear, I mostly find it frustrating that this was hacked into the packages by MS, and means you can't trust anything about supported frameworks on Nuget any more 😞

**Comment by nblumhardt** [comment]
Follow up on "option 2" from: https://github.com/serilog/serilog/issues/1970#issuecomment-1813118532.

We definitely could conditionally exclude the dependency on _System.Diagnostics.DiagnosticSource_ in the `netstandard2.0` build configuration, and `#if` out the relevant parts of the codebase.

I'd dearly like to avoid that if it's feasible. Serilog just emerged from a long period of slow change in which we accumulated a lot of these, and in 3.0 we managed to clean a good number of them out. I'd like to clean out more: working on code that's full of conditional compilation isn't fun: automated refactorings don't always work, algorithms are hard to follow, and dubious bloat accumulates.

If we end up wanting to ship a 3.1 patch that works on .NET 2.1 and 3.0, I think the way to do it would be:

 1. Create a release branch `release/3.1` from `dev`
 2. Make the changes directly in `release/3.1`
 3. Ship the patch directly from `release/3.1` (modify the build so that release branches are published - don't merge to `main`)
 4. Immediately bump `dev` up to 4.0
 5. Backport any remaining bug fixes to 3.1 as required, until 4.0 ships

It's not an appealing option but there if we absolutely must do it :-)

**Comment by andrewlock** [comment]
IMO, it's not worth the hassle. Anyone on 2.1/3.0 is way behind the curve at this point, so I personally think it's ok to chalk this one up to experience, and bear it in mind in any other package updates. At the very least, wait maybe until someone _genuinely_ screams 😄 

I'm just trying to make more people aware of the issue with the MS packages in general tbh, so people consider it when they're adding/bumping dependencies 🙂

**Comment by nblumhardt** [comment]
Thanks again for the heads-up and all the input 👍
